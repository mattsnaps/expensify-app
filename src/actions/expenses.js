import database from "../firebase/firebase";
import { ref, push, get, remove } from "firebase/database";


// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };
        const databaseRef = ref(database, '/expenses');
        return push(databaseRef, expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id: id } = {}) => ({
    id: id,
    type: 'REMOVE_EXPENSE'
});

export const startRemoveExpense = ({ id: id } = {}) => {
    return (dispatch) => {
        const databaseRef = ref(database, `expenses/${id}`);
        return remove(databaseRef).then(() => {
            dispatch(removeExpense({
                id
            }));
        });
    };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

//SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        const expenseList = [];
        const databaseRef = ref(database, '/expenses');
        return get(databaseRef).then((snapshot) => {
            snapshot.forEach((childExpense) => {
                expenseList.push({
                    id: childExpense.key,
                    ...childExpense.val()
                });
            });
            dispatch(setExpenses(expenseList));
        });

    };
};
