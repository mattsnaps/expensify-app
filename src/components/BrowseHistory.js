const BrowseHistory = {
    navigate: null,
    push: (page) => typeof BrowseHistory.navigate === "function" && BrowseHistory.navigate(page)
};

export default BrowseHistory;