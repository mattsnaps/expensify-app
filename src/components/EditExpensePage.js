import React, {useCallback} from "react";
import {MemoryRouter, useHistory, useLocation, useParams} from "react-router-dom";

const EditExpensePage = () => {
    console.log();
    return (
        <div>
            Editing the expense with id of {useParams().id}
        </div>
    );
};

export default EditExpensePage;