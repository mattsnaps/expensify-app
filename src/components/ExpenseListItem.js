import React from "react";

const ExpenseListItem = (props) => (
    <div>
        <h3>{props.description}</h3>
        <p>{props.amount} --- {props.createdAt}</p>
    </div>
);

export const ExpenseListItem2 = ({ description, amount, createdAt }) => (
    <div>
        <h3>{description}</h3>
        <p>{amount} --- {createdAt}</p>
    </div>
);

export default ExpenseListItem ;