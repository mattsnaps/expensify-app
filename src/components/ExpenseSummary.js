import React from "react";
import { connect, useSelector } from "react-redux";
import getVisibleExpenses from "../selectors/expenses";
import selectExpensesTotal from "../selectors/expenses-total";
import numeral from "numeral";

const ExpenseSummary = () => {

    const expenses = useSelector((state) => getVisibleExpenses(state.expenses, state.filters));
    const expenseCount = expenses.length;
    const total = selectExpensesTotal(expenses);

    return(
        <div>
            <p>Viewing {expenseCount} expenses for a total of: {numeral(total / 100).format('$0,0.00')}</p>
        </div>
    );
};

export default connect()(ExpenseSummary);