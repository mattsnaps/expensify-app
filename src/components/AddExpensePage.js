import React from "react";
import { connect } from 'react-redux';
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";
import { useNavigate } from "react-router-dom";

export const AddExpensePage = (props) => {

    const navigate = useNavigate();

    const onSubmit = (expense) => {
        props.startAddExpense(expense);
    };

    return (
        <div>
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Add Expense</h1>
                </div>
            </div>
            <div className="content-container">
                <ExpenseForm
                    onSubmit2={
                        (expense) => {
                            onSubmit(expense);
                            navigate('/', {replace: true});
                        }
                    }
                />
            </div>

        </div>
    );
};



const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);