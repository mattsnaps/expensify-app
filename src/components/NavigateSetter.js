import { useNavigate } from 'react-router-dom';
import BrowseHistory from "./BrowseHistory";

const NavigateSetter = () => {
    BrowseHistory.navigate = useNavigate();
    console.log(1, BrowseHistory);

    return null;
};

export default NavigateSetter;