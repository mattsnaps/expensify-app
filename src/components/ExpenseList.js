import React from "react";
import { connect } from "react-redux";
import ExpenseListItem, { ExpenseListItem2 } from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {
            props.expenses.map((item) => (
                <ExpenseListItem
                    key={item.id}
                    description={item.description}
                    amount={item.amount}
                    createdAt={item.createdAt}
                />
            ))
        }

        <h1>Expense List v.2</h1>
        {
            props.expenses.map((item) => (
                <ExpenseListItem2 key={item.id} {...item} />
            ))
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);