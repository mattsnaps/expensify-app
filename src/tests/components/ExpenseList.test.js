import React from "react";
import { shallow } from "enzyme";
import { ExpenseList } from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";

test("ExpenseList Render expense list with expenses Success", () => {
   const wrapper = shallow(<ExpenseList expenses={expenses}/>);
   expect(wrapper).toMatchSnapshot();
});

test("ExpenseList Render expense list with no expenses Success", () => {
   const wrapper = shallow(<ExpenseList expenses={[]}/>);
   expect(wrapper).toMatchSnapshot();
});
