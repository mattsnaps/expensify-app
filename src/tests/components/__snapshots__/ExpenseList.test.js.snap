// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseList Render expense list with no expenses Success 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;

exports[`ExpenseList Render expenselist with expenses Success 1`] = `
<div>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Market"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={1095}
    createdAt={-345600000}
    description="Rent Card"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={4050}
    createdAt={345600000}
    description="Credit Card"
    id="3"
    key="3"
    note=""
  />
</div>
`;
